src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+namespace Microsoft.Rest
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+
+    /// <summary>
+    /// Abstract base class for types that act like enums, but can be extended with arbitrary string values.
+    /// </summary>
+    public abstract class ExtensibleEnum<T> : IEquatable<T> where T : ExtensibleEnum<T>
+    {
+        private static readonly Lazy<Dictionary<string, T>> _nameMap =
+            new Lazy<Dictionary<string, T>>(CreateNameMap, isThreadSafe: true);
+
+        private string _name;
+
+        /// <summary>
+        /// Initializes a new instance of the ExtensibleEnum class.
+        /// </summary>
+        /// <param name="name">The value for this instance of the extensible enumeration.</param>
+        protected ExtensibleEnum(string name)
+        {
+            _name = name;
```

Comment:
populate the `Dictionary` right here, also don't make it `Lazy`

Reply:
Populate the static dictionary here? Don't get you....

Reply:
@olydis Any particular reason `Lazy` is not a good idea here? My intent was that code that doesn't use any of the extensible enums won't have to pay the cost of reflection for building the map.

Reply:
agree about the reflection, actually the plan is to get rid of that as well - as far as I'm concerned, the Dictionary should map *raw values* to *wrapped values*, i.e. the keys don't really matter. Thinking about the scenarios:

say we have enum `Color` mapping pascal-cased color names, because C# to lower case color names for the wire
- if you wanna use a "well-known" enum value, you do so using `Color.Red`
- if the color you want is not there (but you really know that the server accepts it), you gotta do `Color.Create('cyan')`, i.e. provide the wire value *anyways* (or should the Create function know that it must lowercase stuff? even worse more so if it's an int enum)

So actually, I was somewhat confused about the behavior before. Also imagine the deserializer encounters something it hasn't seen before, won't it use `Create` using the *value* to create the new wrapped instance? So even if it sees `red`, how does it know to return *the very* `Color.Red` if the Dictionary is a mapping from name to value, i.e. `Red` to the wrapped value?

Reply:
@olydis 
> agree about the reflection, actually the plan is to get rid of that as well 

Did you mean get rid of the dictionary entirely, or just change how it works...? I don't think you can do either without losing the "interning" behavior.

> as far as I'm concerned, the Dictionary should map raw values to wrapped values, i.e. the keys don't really matter.

It already does this. Note the call to `GetValue` here:

```csharp
IEnumerable<T> allKnownValues = 
                allPublicStaticFields 
                    .Where(f => f.FieldType == typeof(T)) 
                    .Select(f => f.GetValue(null)) 
                    .OfType<T>(); 
```

When it gets converted to a dictionary, the key comes from the explicit `operator string`, which calls `ToString`, which returns `_name`. This is admittedly confusing... The field should have been called `_value` instead of `_name`, and the code to build the dictionary should have looked like this:

```csharp
 return allKnownValues.ToDictionary(v => v._value, v => v); 
```

But I think it already does what you want it to do.

> if the color you want is not there (but you really know that the server accepts it), you gotta do Color.Create('cyan'), i.e. provide the wire value anyways (or should the Create function know that it must lowercase stuff? even worse more so if it's an int enum)

As originally designed, you'd have to provide the wire value, which ideally would be documented in the REST API reference. `Create` is not meant to be "smart" in any way -- it just looks up exactly the key you give it, and creates a new wrapper if it isn't found. This should give the most predictable behavior.

Note that I originally had an implicit conversion from `ExtensibleEnum` to `string` (read: `TVal`), but that turned out to be very problematic in practice. Therefore, the intent is for all usage to be in terms of the `ExtensibleEnum`-derived type, unless calling code explicitly wants to deal with `TVal`s (or needs to in case of using new server features from an old client).

Reply:
@olydis any thoughts about why making the dictionary Lazy is not a good idea here?